### PR TITLE
Package dependencies for ble2mqtt python package

### PR DESCRIPTION
**Pull request by QuintinHill**
Maintainer: me 
Compile tested: Unfortunately not (I tried without success to build this out on my machine but I think this is more of a failure in my setup.)  Hopefully CI will test this.
Run tested: Unfortunately not

Description: [ble2mqtt](https://github.com/devbis/ble2mqtt) is an interesting python package unfortunately it requires [Bleak](https://github.com/hbldh/bleak) which requires [dbus-fast](https://github.com/Bluetooth-Devices/dbus-fast) and this latter package cannot be successfully installed on a router with 128MB of RAM using pip (it runs out of memory and crashes).  This PR packages the dependencies for ble2mqtt but dbus-fast is the priority here.

### CODE REVIEW

File: lang/python/python-ble2mqtt/Makefile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+define Package/python3-ble2mqtt
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Bluetooth to MQTT bridge
+  URL:=https://github.com/devbis/ble2mqtt/
+  DEPENDS:=+python3 \
+  	+python-bleak
+  	+python-aio-mqtt-mod
+  VARIANT:=python3
```
Comment (1715173329) [inline lang/python/python-ble2mqtt/Makefile:33]:
```suggestion
  DEPENDS:=+python3 +python-bleak +python-aio-mqtt-mod
```

the dependency list is short enough, no need to wrap new lines.
`VARIANT:=python3` has been deprecated in py3 makefile.

File: lang/python/python-bleak/Makefile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+define Package/python3-bleak
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:= A cross platform Bluetooth Low Energy Client for Python using asyncio
+  URL:=https://github.com/hbldh/bleak
+  DEPENDS:=+python3 \
+  	+python3-asyncio
+	+python-dbus-fast
+  VARIANT:=python3
```
Comment (1715173329) [inline lang/python/python-bleak/Makefile:33]:
```suggestion
  DEPENDS:=+python3 +python3-asyncio +python-dbus-fast
```

File: lang/python/python-dbus-fast/Makefile [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+define Package/python3-dbus-fast
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Python DBus library
+  URL:=https://github.com/Bluetooth-Devices/dbus-fast
+  DEPENDS:=+python3 \
+  	+python3-async-timeout
+  VARIANT:=python3
```
Comment (1715173329) [inline lang/python/python-dbus-fast/Makefile:32]:
```suggestion
  DEPENDS:=+python3 +python3-async-timeout
```

File: lang/python/python-aio-mqtt-mod/Makefile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+include ../python3-package.mk
+
+define Package/python3-aio-mqtt-mod
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Python asynchronous MQTT client for 3.1.1 protocol version.
+  URL:=https://github.com/devbis/aio-mqtt
+  DEPENDS:=+python3
+  VARIANT:=python3
```
Comment (1715173329) [inline lang/python/python-aio-mqtt-mod/Makefile:31]:
```suggestion
```

**Review by 1715173329 — CHANGES_REQUESTED** [review]

File: lang/python/python-dbus-fast/Makefile [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+include ../pypi.mk
+include $(INCLUDE_DIR)/package.mk
+include ../python3-package.mk
+
+define Package/python3-dbus-fast
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Python DBus library
```
Comment (jefferyto) [inline lang/python/python-dbus-fast/Makefile:28]:
```suggestion
  TITLE:=A faster version of dbus-next
```

**Review by jefferyto — COMMENTED** [review]
I also suggest you:

* Test build these packages using a local build environment. I suggest [using the SDK](https://openwrt.org/docs/guide-developer/toolchain/using_the_sdk).

  If you do this (after applying the changes suggested by @1715173329 and myself) you will find the packages will still not build, for reasons outside of your control, but this is still important because:

* Run, for each package:

  ```
  make package/<package name>/{clean,configure} PY3=stdlib V=s
  ```

  This will show you which Python standard library packages are used by each package. If a package requires 3 or less standard library packages, I suggest replacing the `+python3` dependency with `+python3-light` plus the reported dependencies.

File: lang/python/python-bleak/Makefile [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+include ../python3-package.mk
+
+define Package/python3-bleak
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:= A cross platform Bluetooth Low Energy Client for Python using asyncio
+  URL:=https://github.com/hbldh/bleak
+  DEPENDS:=+python3 \
+  	+python3-asyncio
```
[the remark this answers is not in the document]
Comment (karlp) [reply lang/python/python-bleak/Makefile:31]: well, yes, but also no.  You absolutely can't use bleak withotu asyncio, but that's stricly speaking on the user end.
I'd packaged this locally, but was batching it up with some other stuff.  I just used python3-light, and that's "enough" but you absolutely need asyncio to _use_ bleak.

**Comment by karlp** [comment]
The bleak and dbus-fast packages (after suggested changes), are virtually identical to what I've been using locally, so they look good.  No comments on the other two

File: lang/python/python-bleak/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (QuintinHill) [reply lang/python/python-bleak/Makefile]: I've made it depend on both and put a note about the runtime dependency in the commit message.

**Comment by QuintinHill** [comment]
@jefferyto  I tried those commands but got no error apart from:
`WARNING: Makefile 'package/feeds/local/python-ble2mqtt/Makefile' has a dependency on 'python-bleak', which does not exist
WARNING: Makefile 'package/feeds/local/python-ble2mqtt/Makefile' has a dependency on 'python-aio-mqtt-mod', which does not exist
WARNING: Makefile 'package/feeds/local/python-bleak/Makefile' has a dependency on 'python-dbus-fast', which does not exist
`
It was a useful exercise as I found had got the PyPI name of dbus-fast wrong (it needs to be dbus_fast).  Hopefully everything is more in order now.  Thank you for your help with this.

File: lang/python/python-bleak/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (QuintinHill) [reply lang/python/python-bleak/Makefile]: It looks like in release of v0.20.0 release of bleak a few minutes ago the dependency will be dropped when using python 3.11+

File: lang/python/python-ble2mqtt/Makefile [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+include $(INCLUDE_DIR)/package.mk
+include ../python3-package.mk
+
+define Package/python3-ble2mqtt
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Bluetooth to MQTT bridge
+  URL:=https://github.com/devbis/ble2mqtt/
+  DEPENDS:=+python3-light +python-aio-mqtt-mod +python3-asyncio +python-bleak +python3-logging
```
Comment (jefferyto) [inline lang/python/python-ble2mqtt/Makefile:30]:
```suggestion
  DEPENDS:=+python3-light +python3-aio-mqtt-mod +python3-asyncio +python3-bleak +python3-logging
```

File: lang/python/python-bleak/Makefile [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+include ../python3-package.mk
+
+define Package/python3-bleak
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Bluetooth Low Energy platform Agnostic Klient
+  URL:=https://github.com/hbldh/bleakpython3-ctypes
+  DEPENDS:=+python3-light +python3-async-timeout +python3-asyncio \
+	+python-dbus-fast +python3-ctypes +python3-logging
```
Comment (jefferyto) [inline lang/python/python-bleak/Makefile:31]:
```suggestion
	+python3-dbus-fast +python3-ctypes +python3-logging
```

File: lang/python/python-dbus-fast/Makefile [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+include ../python3-package.mk
+
+define Package/python3-dbus-fast
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=A faster version of dbus-next
+  URL:=https://github.com/Bluetooth-Devices/dbus-fast
+  DEPENDS:=+python3-light +python3-async-timeout +python3-asyncio \
+	+python3-distutils +python3-logging +python3-urllib +python3-xml
```
Comment (jefferyto) [inline lang/python/python-dbus-fast/Makefile:31]:
```suggestion
	+python3-logging +python3-urllib +python3-xml
```

distutils is only used in [build_ext.py](https://github.com/Bluetooth-Devices/dbus-fast/blob/v1.84.2/build_ext.py#L4), which should be only used during build/setup and not run-time.

File: lang/python/python-dbus-fast/Makefile [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=python-dbus-fast
+PKG_VERSION:=1.84.2
+PKG_RELEASE:=1
+
+PYPI_NAME:=dbus_fast
```
Comment (jefferyto) [inline lang/python/python-dbus-fast/Makefile:13]:
```suggestion
PYPI_NAME:=dbus-fast
PYPI_SOURCE_NAME:=dbus_fast
```

**Comment by jefferyto** [comment]
> I did see the following warnings but I guess they can be ignored:

The names in `DEPENDS` should be the opkg package names (e.g. `python3-bleak`) not the source package name (python-bleak).

Also, since bleak is a new package, I would prefer if the "upgrade" commit was squashed into the "add package" commit.

File: lang/python/python-bleak/Makefile [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+include ../pypi.mk
+include $(INCLUDE_DIR)/package.mk
+include ../python3-package.mk
+
+define Package/python3-bleak
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Bluetooth Low Energy platform Agnostic Klient
+  URL:=https://github.com/hbldh/bleakpython3-ctypes
```
Comment (QuintinHill) [inline lang/python/python-bleak/Makefile:29]: Copy and paste error!

**Comment by QuintinHill** [comment]
I'm not sure why the test is failing:
```
Testing package python3-ble2mqtt-src in version 0.1.5 from python-ble2mqtt
Installing python3-ble2mqtt-src (0.1.5-1) to root...
Configuring python3-ble2mqtt-src.
No test.sh script available
Removing package python3-ble2mqtt-src from root...
Testing package python3-ble2mqtt in version 0.1.5 from python-ble2mqtt
Unknown package 'python3-ble2mqtt'.
Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency python3-aio-mqtt-mod for python3-ble2mqtt
 * pkg_hash_check_unresolved: cannot find dependency python3-bleak for python3-ble2mqtt
 * pkg_hash_fetch_best_installation_candidate: Packages for python3-ble2mqtt found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package python3-ble2mqtt.
Error: Process completed with exit code 255.
```

I have the copy & paste error to fix but I will refrain from updating until I can understand why this is failing.

**Comment by jefferyto** [comment]
> I'm not sure why the test is failing:

That error is because those packages are new, they are not available to download by opkg yet. The run-test installs each package from the locally compiled ipk file, but opkg doesn't know to look to other locally compiled ipk files to satisfy dependencies.

I think this is okay to merge once the bleak URL has been fixed :smile:

**Comment by QuintinHill** [comment]
Done!

**Comment by jefferyto** [comment]
Thanks again!

**Pull request merged**